src/devices/machine/sis630_host.cpp

```diff
@@ -0,0 +1,444 @@
+{
+	LOGIO("Write integrated VGA control data [$9c] %02x\n", data);
+	// TODO: "integrated VGA control" (?)
+	m_vga_control = data;
+//  remap_cb();
+}
+
+u32 sis630_host_device::agp_priority_timer_r(offs_t offset, uint32_t mem_mask)
+{
+	LOGIO("Read AGP priority timer [$a0] %08x (%08x)\n", mem_mask, m_agp_priority_timer);
+	return m_agp_priority_timer;
+}
+
+void sis630_host_device::agp_priority_timer_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	COMBINE_DATA(&m_agp_priority_timer);
+	LOGIO("Write AGP priority timer [$a0] %08x & %08x (%08x)\n", data, mem_mask, m_agp_priority_timer);
+}
+
+u8 sis630_host_device::agp_mailbox_r(offs_t offset)
+{
+	LOGIO("Read AGP mailbox [$%02x] (%02x)\n", offset + 0xa4, m_agp_mailbox[offset]);
+	return m_agp_mailbox[offset];
+}
+
+void sis630_host_device::agp_mailbox_w(offs_t offset, u8 data)
+{
+	LOGIO("Write AGP mailbox [$%02x] %02x\n", offset + 0xa4, data);
+	m_agp_mailbox[offset] = data;
+}
+
+// TODO: move to generic interface
+u32 sis630_host_device::agp_id_r()
+{
+	LOGAGP("Read AGP ID [$c0]\n");
+	// bits 23-16 AGP v2.0
+	// bits 15-8 0x00 no NEXT_PTR (NULL terminates here)
+	// bits 7-0 CAP_ID (0x02 for AGP)
+	return 0x00200002;
+}
+
+u32 sis630_host_device::agp_status_r()
+{
+	LOGAGP("Read AGP status [$c4]\n");
+	// bits 31-24 RQ max number of AGP command requests (0x1f + 1 = 32)
+	// bit 9: SBA, side band addressing enabled
+	// ---- -xxx RATE
+	// ---- -1-- 4X transfer capable
+	// ---- --1- 2X transfer capable
+	// ---- ---1 1X transfer capable
+	// NB: documentation claims a RATE of 0x03 then contradicts with "111b" value, do the math
+	// It gets setup with a 4X at POST, assume 0x07 is right
+
+	// Stuff that isn't enabled here:
+	// bit 5: 4G support address greater than 4 GB
+	// bit 4: FW transfer support
+
+	return 0x1f000207;
+}
+
+u32 sis630_host_device::agp_command_r(offs_t offset, uint32_t mem_mask)
+{
+	LOGAGP("Read AGP command [$c8] %d %d %02x\n", m_agp.sba_enable, m_agp.enable, m_agp.data_rate);
+	// TODO: enable gets cleared by AGP_RESET, or even from PCI RST#
+	return m_agp.sba_enable << 9 | m_agp.enable << 8 | (m_agp.data_rate & 7);
+}
+
+void sis630_host_device::agp_command_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	LOGAGP("Write AGP command [$c8] %08x & %08x\n", data, mem_mask);
+	if (ACCESSING_BITS_8_15)
+	{
+		m_agp.sba_enable = bool(BIT(m_agp.sba_enable, 9));
+		m_agp.enable = bool(BIT(m_agp.enable, 8));
+		LOGAGP("- SBA_ENABLE = %d AGP_ENABLE = %d\n", m_agp.sba_enable, m_agp.enable);
+	}
+
+	if (ACCESSING_BITS_0_7)
+	{
+		std::map<u8, std::string> agp_transfer_rates = {
+			{ 1, "1X" },
+			{ 2, "2X" },
+			{ 4, "4X" }
+		};
+
+		// make sure the AGP DATA_RATE specs are honored
+		try {
+			const u8 data_rate = data & 7;
+			LOGAGP("- DATA_RATE = %s\n", agp_transfer_rates.at(data_rate));
+			m_agp.data_rate = data_rate;
+		}
+		catch (std::out_of_range& err) {
+			LOG("Warning: AGP illegal DATA_RATE set = %d enabled=%d\n", data & 7, m_agp.enable);
+		}
```

Comment:
Once again, it’s bad form to do stuff in a way that requires allocations and exceptions in a handler like this.